### DISCUSSION TITLE
IamForce2 on MPC LIVE 2

### DISCUSSION THREAD

**Discussion by TheKikGen** [TheKikGen/MPCLiveXplore-libs#9] [General]
IamForce allows your MPC to become a Force. IamForce2 is an entirely rewritten version of the first version, and now supports several midi devices with a plugin system. A new standalone mode is also available if you haven't any external controller.

IamForce2 does not update your MPC, but prepares and launches an isolated environment for the Force software ONLY from the external SD card.

**Comment by TheKikGen** [comment]
On the MPC Live 2, the order of card is something different than the MPC Live I use for development.

```
root@mpc-live-ii:/# amidi -l
Dir Device Name
IO hw:1,0,0 MPC Public
IO hw:1,0,1 MPC Private
IO hw:1,0,2 MPC MIDI Port A
IO hw:1,0,3 MPC MIDI Port

```
![image](https://github.com/TheKikGen/MPCLiveXplore-libs/assets/35703467/9376e53e-3bed-493d-a902-f9123d2223e3)


I confirm the bug in tkgl_midimapper (the base of IamForce2). It takes the audio card and not the midi card in your case , because it is the first to match....a patch is currently tested.

**Reply by TheKikGen** [reply]
The "GetCardFromShortName" function did not test the Alsa card type.  This can be done with the ctl api.  The new function looks like :

```
///////////////////////////////////////////////////////////////////////////////
// Get an ALSA card from a matching regular expression pattern
///////////////////////////////////////////////////////////////////////////////
int GetCardFromShortName(const char *pattern) {
   int card = -1;
   int device = -1;
   snd_ctl_t *ctl;
   char name[32];
   char* shortname = NULL;


   if ( snd_card_next(&card) < 0) return -1;
   while (card >= 0) {
   	if ( snd_card_get_name(card, &shortname) == 0  ) {
       if (match(shortname,pattern) ) {
	        sprintf(name, "hw:%d", card);
	        if ( snd_ctl_open(&ctl, name, 0) < 0) {
                tklog_error("Impossible to open control for card %d - %s (GetCardFromShortName).\n",card,shortname);
		        return -1 ;
            }
            device = -1 ;
    		if ( snd_ctl_rawmidi_next_device(ctl, &device) >= 0 ) {
                snd_ctl_close(ctl) ;                 
                return card;
		    }
            snd_ctl_close(ctl) ;                 
	    }
    }
   	if ( snd_card_next(&card) < 0) break;
   }

   return -1;
}
```
If the card is an audio card, then no midi device exists, and the next card will be read.

**Reply by int3ks** [reply]
is the new code allready included in the version you posted on facebook?

**Reply by int3ks** [reply]
![image](https://github.com/TheKikGen/MPCLiveXplore-libs/assets/22541283/ed59cfa3-71b1-4b06-95ee-56bb8878d350)

**Reply by int3ks** [reply]
it still trys to open "hw:0,0,1" but there is no device with 0,0,1 ??

**Reply by int3ks** [reply]
can you add more debug (tklog_info) output? in GetCardFromShortName and GetSeqClientFromRegEx .... maybe it help

**Reply by TheKikGen** [reply]
Try this one. I added a correction + more debug log msg.
[tkgl_midimapper.zip](https://github.com/TheKikGen/MPCLiveXplore-libs/files/13313431/tkgl_midimapper.zip)

**Reply by int3ks** [reply]
it starts now :) good job ! thx
![image](https://github.com/TheKikGen/MPCLiveXplore-libs/assets/22541283/66009fa1-1527-497d-af63-271cad93c65d)

**Reply by int3ks** [reply]
some of the buttons on my MPC Live II stays dark and have no function...  is that intentional? Can I somehow integrate the midi.cfg file from Hakai mod?

**Reply by int3ks** [reply]
[midi.cfg.zip](https://github.com/TheKikGen/MPCLiveXplore-libs/files/13317102/midi.cfg.zip)

**Reply by TheKikGen** [reply]
IamForce2 is not compatible with Hakai.  Making an external configuration file is an heavy task (I did it with the V1), it is why IamForce is built in a more static approach.  We can, for sure, improve the current MPC mapping if you have any suggestions.  What are the buttons not lighted and what kind of usage are you thinking with ?

**Reply by int3ks** [reply]
ok. i understand... 
pls take a look at this midi.cfg .... the hakai mapping is allready really good and it will help all users switch from hakai to iamforce!
in this file are one side the force_buttons on the other side the mpc_buttons in comments (and the midi code as hex value.)

or think about also supporting this format....  then each hardware plattform can have its own mapping. and its is user customizable and also interchangeable with hakai.... ;)

**Reply by TheKikGen** [reply]
In fact using IamForce only with the MPC pads and buttons is a pain.
I recommend an external controller.  In my opinion, setting a "standard" mapping for MPCs, with maybe some variation between Live, Live2, X, One could be a good point.  Better than customize everything because it is, then , very difficult to maintain.
I don't say that i will propose one day, a config file, but at the moment,  I consider this as a low priority dev.
I concentrate my efforts in code stabilization (with your help !) first.

**Reply by int3ks** [reply]
i have no problem or need for the 8x8 grid.... clips can be launched also with the touchscreen... The most important thing for me is the arranger! and diskstreaming with audiotracks in the arranger..... maybe for the stepsequencer it would be nize to have the 8x8 grid .... 

long story short . I'll make a few notes about which buttons are missing or don't have a function. and which functions make sense on which buttons....

**Reply by int3ks** [reply]
first try.->
Bank buttons should work for pads like on MPC

-the "-"(minus) button switch bank button and PAD to MPC style mode
-the "+"(plus) button switch bank buttons and PADs to Force style mode 
(in force mode the bank button change the quadrant)

#####

Atm the "C/G Bank" button set the StepSeq mode -> should be on StepSEQ button (atm without funktion)
 
Atm "TC" button without funktion -> should switch TC

Atm "ERASE" button akt as Edit button -> should be the erase button

Atm "MIX" button without funktion -> should be mixer/pad mixer  

"FULLLEVEL" can be better used for Arranger/Track edit (which is currently on bank "D/H Bank" button)

the "MUTE"/Sampleedit button should be Sampler/Sampleedit

the "OVERDUB" button is currently the save button -> should be the overdub button

the "PLAY START" button has no funktion atm -> should be maybe FileBrowser or the save button or both with shift?

"NEXTSEQ" XYFX doesnt work with shift it should open XYFX at least . 

#####

qlink knobs.
qlink press switch between 1-8 and 9-16
is it possible to use shift? 
qlink 1-4 with shift => 5-8
qlink 9-12 with shift => 13-16

**Reply by int3ks** [reply]
hi... got my launchpad mini today... would be nize if launchpad connected, the pads on MPC are allways the lower left quadrant. makes no sense if it shows a half row of the stepsequencer part ;)

btw. the stop key is not as expected. it stops on shift+stop and stop stops the arranger play mode ?! should be exactly the opposite

**Reply by int3ks** [reply]
OK. I figured out how to change the pad bank... with minus and the pads 7,8,3,4... as for that, forget what I said... but it would be nice, especially if you have the forcepad in the drum or melodic mode used when the pads on the mpc follow... when I press pad on the launchpad which then switches the mpcpad to the same quadrant

**Reply by int3ks** [reply]
The other buttons/functions are also more a matter of habit... but it would be good if at least the buttons do that, as they are labeled... e.g. "stepseq", "mix" or "overdub" etc

**Reply by kurtjcu** [reply]
For others who may come across this, I had to update(downgrade) via the boostrap image, then update top the latest tkgl ssh image(2.12.2), then put the above file in the modules/iamforcedir. You may be able to do this from the ssh image first, but it will NOT WORK if you use the version of mpc software inside the boostrap image. I just copied via winscp after deleting the original img.

**Reply by TheKikGen** [reply]
FYI, corrected binary of tkgl_midimapper was added to the github.  So the update procedure of the bootstrap will make Iamforce to work with MPCLIVE 2..

**Reply by TheKikGen** [reply]
Bootstrap was also updated accordingly.  Use the update script.

**Comment by int3ks** [comment]


**Reply by Pitrakos** [reply]
Please can you do same for a MPC ONE, in one there is six buttons free already for assign more commands . 
I trayed your file in MPC One and works perfect, only in One we have Track Mute button and is not same than Mute in live and doesn´t works .

Thanks a lot.

**Reply by int3ks** [reply]
my buildscripts and binaries are at alpha state ... i also make a pullrequest for official kikgen repro.
if you like you can pick here ->
https://github.com/int3ks/MPCLiveXplore-libs/tree/main/bin
a bin and test.
should be tmm-IamForce-LPMK3-ONE.so 
in doer_list add TKGL_DOER="iamforce2@LPMK3-ONE"

**Reply by Pitrakos** [reply]
Thanks a lot, I´ll check it.

Regards

**Reply by Pitrakos** [reply]
Hi, works perfect.

Will be great if you can implement more Force options in the extra buttons on One.

Thanks a lot!

**Comment by int3ks** [comment]
> Hi, works perfect.
> 
> Will be great if you can implement more Force options in the extra buttons on One.
> 
> Thanks a lot!

Which button is missing? and which force button? btw. I uploaded new binaries . have you tested the latest version?

**Reply by Pitrakos** [reply]
Thanks a lot, I´ll test it and come back to you with comments.

Another question, is possible to map Qlinks controls to an external pots midi controller?

Regards

**Reply by int3ks** [reply]
pls also check this version -> https://github.com/int3ks/MPCLiveXplore-libs/releases/tag/0.2

most pods on external midi controller are not "endless" and even if they are. i dont have such a keyboard to test with, sry. but maybe you find the "new" implementation of the qlink knobs usefull ;)

**Comment by Mliptak05** [comment]
Does the Akai APC Mini MK1 work on this too ? I know MK2 is supported and they are literally the same thing layout-wise... Sorry if it's a dumb question

**Comment by AJ3I** [comment]
@TheKikGen  @int3ks  Hi , I have Iam force running on Live 2 but no audio through inbuilt speaker, also how can i make use of live2 unused buttons like Mix, StepSEQ etc to map to force functionalities, please guide. Thanks
